src/snowflake/connector/util_text.py

```diff
@@ -7,223 +7,86 @@
-                    col += 1
-        line = buf.readline()
-
-    if len(statement) > 0:
-        yield _concatenate_statements(statement)
-
-
-def _concatenate_statements(statement_list):
-    """Concatenate statements.
-
-    Each statement should be a tuple of statement and is_put_or_get.
-
-    The is_put_or_get is set to True if the statement is PUT or GET otherwise False for valid statement.
-    None is set if the statement is empty or comment only.
-
-    Args:
-        statement_list: List of statement parts.
-
-    Returns:
-        Tuple of statements and whether they are PUT or GET.
+    """For automatic split, sep_sqls(sql) will detect the boundaries of sql statement blocks
+    using a simplified ANTLR4 parser. Or users could use manual way to guide the splitting
+    by using special comment. A pure comment line '--SF:<snowflake splitting hints>'
+    All these commands should not be nested
+      (1) '--SF:auto_split off': disable auto split. This should be first cmd in a script file
+      (2) to disable split the sql texts between these two hints
+      '--SF:<<'
+         ...
+      '--SF:>>'
+      (3) to split the blocks before and after this comment '--SF:<<>>'
     """
-    valid_statement_list = []
-    is_put_or_get = None
-    for text, is_statement in statement_list:
-        valid_statement_list.append(text)
-        if is_put_or_get is None and is_statement and len(text.strip()) >= 3:
-            is_put_or_get = text[:3].upper() in ("PUT", "GET")
-    return "".join(valid_statement_list).strip(), is_put_or_get
+    sqltext = buf.read()
+    sfcmds = []
+    for m in sf_comment_cmd_pattern.finditer(sqltext.lower()):
+        sfcmds.append([m.start(), m.end()])
+
+    # stmts to store the split result to return
+    stmts = []
+
+    is_auto_split = True
+
+    for cmd in sfcmds:
+        if is_auto_split and re.match(
+            r"(\n|\r)+\s*--sf:\s*auto_split\s+off", sqltext[cmd[0] : cmd[1]].lower()
+        ):
+            _logger.info("auto_split off")
+            is_auto_split = False
+    starting_pos_of_block = -1  # not in block yet
+    last_pos = 0  # last pos of ending blocks
+    pieces = []
+    for cmd in sfcmds:
+        if starting_pos_of_block < 0 and re.match(
+            r"(\n|\r)+\s*--sf:\s*<<\s*(\n|\r)+", sqltext[cmd[0] : cmd[1]].lower()
+        ):
+            starting_pos_of_block = cmd[1]
+            pieces.append([0, sqltext[last_pos : (cmd[0])]])
+        elif starting_pos_of_block > 0 and re.match(
+            r"(\n|\r)+\s*--sf:\s*>>\s*(\n|\r)+", sqltext[cmd[0] : cmd[1]].lower()
+        ):
+            pieces.append([1, sqltext[starting_pos_of_block : (cmd[0])]])
+            last_pos = cmd[1]
+            starting_pos_of_block = -1
+        elif starting_pos_of_block < 0 and re.match(
+            r"(\n|\r)+\s*--sf:\s*<<>>\s*(\n|\r)+", sqltext[cmd[0] : cmd[1]].lower()
+        ):
+            pieces.append([0, sqltext[last_pos : (cmd[0])]])
+            last_pos = cmd[1]
```

Comment:
Also since these Regexes are defined in a for loop, we should really compile them outside to only compile them once.

Sorry, I forgot to add this when I was reviewing your code and just remembered it.